lib/PuppeteerSharp/WaitTask.cs

```diff
@@ -1,17 +1,192 @@
+        fulfill();
+      }
+      const success = predicate.apply(null, args);
+      if (success) {
+        observer.disconnect();
+        fulfill(success);
+      }
+    });
+    observer.observe(document, {
+      childList: true,
+      subtree: true,
+      attributes: true
+    });
+    return result;
+  }
+
+  /**
+   * @return {!Promise<*>}
+   */
+  function pollRaf() {
+    let fulfill;
+    const result = new Promise(x => fulfill = x);
+    onRaf();
+    return result;
+
+    function onRaf() {
+      if (timedOut) {
+        fulfill();
+        return;
+      }
+      const success = predicate.apply(null, args);
+      if (success)
+        fulfill(success);
+      else
+        requestAnimationFrame(onRaf);
+    }
+  }
+
+  /**
+   * @param {number} pollInterval
+   * @return {!Promise<*>}
+   */
+  function pollInterval(pollInterval) {
+    let fulfill;
+    const result = new Promise(x => fulfill = x);
+    onTimeout();
+    return result;
+
+    function onTimeout() {
+      if (timedOut) {
+        fulfill();
+        return;
+      }
+      const success = predicate.apply(null, args);
+      if (success)
+        fulfill(success);
+      else
+        setTimeout(onTimeout, pollInterval);
+    }
+  }
+}";
+
+        internal WaitTask(Frame frame, string predicateBody, string polling, int timeout, object[] args)
+        {
+            _frame = frame;
+            _predicateBody = $"return ( {predicateBody} )(...args)"; //predicateBody;
+            _polling = polling;
+            _timeout = timeout;
+            _args = args;
+
+            frame.WaitTasks.Add(this);
+            _taskCompletion = new TaskCompletionSource<JSHandle>();
+
+            _cts = new CancellationTokenSource();
+
+            _timeoutTimer = Task.Delay(timeout, _cts.Token).ContinueWith(_
+                => Termiante(new PuppeteerException($"waiting failed: timeout {timeout}ms exceeded")));
+
+            Rerun();
+        }
+
+        internal Task Task => _taskCompletion.Task;
+
+        internal async void Rerun()
         {
-            throw new NotImplementedException();
+            var runCount = ++_runCount;
+            JSHandle success = null;
+            Exception exception = null;
+
+            var context = await _frame.GetExecutionContextAsync();
+            try
+            {
+                success = await context.EvaluateFunctionHandleAsync(WaitForPredicatePageFunction, new object[] { _predicateBody, _polling, _timeout }.Concat(_args).ToArray());
+            }
+            catch (Exception ex)
+            {
+                exception = ex;
+            }
+
+            if (_terminated || runCount != _runCount)
+            {
+                if (success != null) await success.Dispose();
+                return;
+            }
+            if (exception != null && await _frame.EvaluateFunctionAsync<bool>("s => !s", success))
+            {
+                if (success != null) await success.Dispose();
```

Comment:
we can't `await success?.Dispose();` since https://github.com/dotnet/csharplang/issues/35